src/actions.js

```diff
@@ -248,21 +257,118 @@ export function validate (
     const result = validateValue(formValue, renderModel)
 
     const promises = []
+    dispatch({
+      type: IS_VALIDATING,
+      isValidating: true
+    })
     validators.forEach((validator) => {
-      promises.push(validator(formValue))
+      const type = typeof validator
+      if (type === 'function') {
+        // Original validation. Always validates
+        promises.push(validator(formValue))
+      }
     })
 
+    const fieldValidationPromises = fieldValidators ? fieldValidation(dispatch, getState,
+      fieldValidators, formValue, initialFormValue, previousValidations) : []
+
     // Promise.all fails in Node when promises array is empty
     if (promises.length === 0) {
-      dispatchUpdatedResults(dispatch, [result])
+      dispatchUpdatedResults(dispatch, [result], getState)
+      dispatchDoneValidating(dispatch, fieldValidationPromises, all)
       return
     }
 
     all(promises)
       .then((snapshots) => {
         const results = _.map(snapshots, 'value')
         results.push(result)
-        dispatchUpdatedResults(dispatch, results)
+        dispatchUpdatedResults(dispatch, results, getState)
+      }).then(() => dispatchDoneValidating(dispatch, fieldValidationPromises, all))
+  }
+}
+function dispatchDoneValidating (dispatch, fieldValidationPromises, all) {
+  if (fieldValidationPromises.length === 0) {
+    dispatch({
+      type: IS_VALIDATING,
+      isValidating: false
+    })
+  } else {
+    return all(fieldValidationPromises).then(() => {
+      dispatch({
+        type: IS_VALIDATING,
+        isValidating: false
       })
+    })
   }
 }
+
+function fieldValidation (dispatch, getState, fieldValidators, formValue, initialFormValue, previousValidations) {
+  let fieldsBeingValidated = []
+  const promises = []
+  fieldValidators.forEach(validator => {
+            /**
+         * Field validation. Only validate if field has changed.
+         * Meant for expensive validations like server side validation
+         */
+
+    /**
+     * Field validator definition
+     * @property {String} field field to validate
+     * @property {String[]} fields fields to validate
+     * @property {Function} validator valdation function to validate field/fields. Must return field within error/warning
+     * @property {Function[]} validators valdation functions to validate field/fields. Must return field within error/warning
+     */
+    const {field, fields, validator: validatorFunc, validators: validatorFuncs} = validator
+
+    const fieldsToValidate = fields || [field]
+    fieldsBeingValidated = fieldsBeingValidated.concat(fieldsToValidate)
+    fieldsToValidate.forEach((field) => {
+      const newValue = _.get(formValue, field)
+      const oldValue = _.get(initialFormValue, field)
+
+          // Check if field value has changed
+      if (!_.isEqual(newValue, oldValue) || !initialFormValue) {
```

Comment:
`_.isEqual` can become slow in the worst case. I'm fine with it now because there are other places in Bunsen that do that but I was curious if you've tried looking at the `valueChangeSet` that is in the store. It holds the actual diff from the last value change where the key is the `bunsenId`.

Reply:
I didn't know about that. I'll take a look

Reply:
@sophypal Don't think I can use `valueChangeSet` due to the way it handles arrays.
![image](https://user-images.githubusercontent.com/9057680/54643499-6bae8180-4a6d-11e9-80fd-a39455241bce.png)

 It would also limit what the use could consider a field. If they want to validate against an object field they should be able to (although makes more sense to against a primitive type). Also limits they can't validate against an array (or specific index's value)

Reply:
Ok. Thanks for taking the time to look.